refactor project model: improve ruby dialect and remove unused code.

vpereira opened the pull request:
moved some code to constant and removed redundant

bgeuken commented on src/api/app/models/project.rb:
```diff
@@ -9,6 +9,8 @@ class Project < ApplicationRecord
   include HasAttributes
   include MaintenanceHelper
 
+  PROJECT_TYPES = ['standard', 'maintenance', 'maintenance_incident',
```
I would call this TYPES. So outside of project class it can be referenced viaPROJECT::TYPES  :-)

bgeuken commented on src/api/app/models/project.rb:
```diff
@@ -234,11 +237,9 @@ def cleanup_before_destroy
-      end
+    packages.where.not(develpackage: { id: nil }) do |pkg|
+      pkg.develpackage_id = nil
+      pkg.save
```
Not 100% sure but I think you could also just update all packages here without any performance loss. Might be worth to check

coolo replied:
As the packages are destroyed 3 lines down - what is the point anyway?

vpereira replied:
yup and in the package model, we do all the trick being done there:

```
before_destroy :delete_on_backend
  before_destroy :close_requests
  before_destroy :update_project_for_product
  before_destroy :remove_linked_packages
  before_destroy :remove_devel_packages
```
https://github.com/vpereira/open-build-service/blob/refactor_project/src/api/app/models/package.rb#L87

hennevogel commented on src/api/app/models/project.rb:
```diff
@@ -215,11 +218,9 @@ def config
-    end
+    deleted_title = 'Place holder for a deleted project instance'
+    project = Project.find_or_create_by(name: 'deleted') { |prj| prj.title = deleted_title }
+    project.store
```
jupp

dmarcoux commented on src/api/app/models/project.rb:
```diff
@@ -745,10 +736,16 @@ def flag_status(default, repo, arch, prj_flags, pkg_flags)
     opts[:explicit] = '1' if expl
-    ret = 'enable' if ret == :enabled
-    ret = 'disable' if ret == :disabled
+    ret_str = case ret
```
The case could be removed since `flags.status` is [an enum](https://github.com/openSUSE/open-build-service/blob/master/src/api/db/structure.sql#L503). The array below could then be `[ret.to_s, opts]`

coolo replied:
well, minus the d at the end.

dmarcoux replied:
Oops. Still the `else` in the case will never be reached.

coolo replied:
sounds like an elvis is overdue :)

vpereira replied:
the end is necessary. I tried to refactor it to if else condition and it broke some test cases

coolo replied:
that possibly means ret is already a string - passed as default

bgeuken commented on src/api/spec/models/project_spec.rb:
```diff
@@ -585,4 +585,17 @@ def reset_project_in_backend
     end
   end
+
+  describe 'foo' do
```
^Looks like a leftover:laughing:

vpereira replied:
I was like Michael Douglas @ Falling down :P 
couldnt test because of the issue with createrepo_c, then merged an approved PR and the whole docker rake tasks are gone and I couldnt test my code anymore.. I was like https://www.youtube.com/watch?v=zJs9p-VNORw :P

dmarcoux commented on src/api/app/models/project/exceptions.rb:
```diff
@@ -0,0 +1,25 @@
+module Project::Exceptions
+  extend ActiveSupport::Concern
+  class CycleError < APIException
```
Let's rename those `SomethingError` classes to `SomethingException` since they inherit from `APIException`

vpereira replied:
I agree with that, ```APIException``` and ```CycleError``` don't match. However looks like in ruby exceptions are named with the suffix Error https://ruby-doc.org/core-2.2.0/Exception.html

So shouldn't we rename ```APIException``` to ```ApiError``` ?

dmarcoux replied:
`APIError` is also fine for me.

ChrisBr replied:
It inherited from Exception some time ago. We changed that [here](https://github.com/openSUSE/open-build-service/commit/2316f5b24f1e16fa02c1b227850ce2e61bb938a0). So the name does not match anymore. Yes rename it!

vpereira commented:
don't be afraid: the amount of files changed, is related to renaming ```APIException to APIError```

vpereira commented:
ready to review. Things that I did:

- added some missing tests
- reviewed project model: removed unused code and improved ruby style
- moved the whole exception class definitions to its own file
- renamed the APIException to APIError as discussed above.

Please review @openSUSE/open-build-service

dmarcoux commented on src/api/app/models/project/exceptions.rb:
```diff
@@ -0,0 +1,24 @@
+module Project::Exceptions
```
Now that this file contains errors, it should be renamed to `errors.rb`.

Could we also have a line between each error class?

I really like this idea. We should do it for other models in other PRs!

dmarcoux commented on src/api/spec/controllers/webui/project_controller_spec.rb:
```diff
@@ -1126,17 +1126,17 @@
     end
 
-    shared_examples 'a non APIException' do |exception_class|
+    shared_examples 'a non APIError' do |exception_class|
```
`exception_class` should now be `error_class`

dmarcoux commented on src/api/spec/controllers/webui/project_controller_spec.rb:
```diff
@@ -1156,7 +1156,7 @@
        BsRequestAction::VersionReleaseDiffers,
        BsRequestAction::UnknownTargetProject,
        BsRequestAction::UnknownTargetPackage].each do |exception_class|
```
Same error, `error_class`

bgeuken commented on src/api/app/models/project/update_from_xml_command.rb:
```diff
@@ -1,5 +1,6 @@
 class Project
   class UpdateFromXmlCommand
+    include Project::Exceptions
```
Doesn't this belong to the previous commit? :thinking:

bgeuken commented on src/api/app/controllers/application_controller.rb:
```diff
@@ -213,7 +213,7 @@ def pass_to_backend(path = nil)
   end
 
-  rescue_from APIException do |exception|
+  rescue_from APIError do |exception|
```
Please add an explanation why we change inherited exception class to the commit message.

vpereira replied:
that's what I don't like from this review system from github. We "decided" to rename it here, in a discussion between @dmarcoux @coolo @ChrisBr and me. But some how it gets collapsed in a way that I cannot find or link it (at least easily) :/

coolo replied:
https://github.com/openSUSE/open-build-service/pull/5531#discussion_r208499449 - I was not involved! :)

vpereira replied:
hahaha you are omnipresent @coolo :)

bgeuken commented on src/api/app/models/project/errors.rb:
```diff
@@ -1,22 +1,29 @@
-module Project::Exceptions
+module Project::Errors
```
Please squash this commit into the one that introduced this file

bgeuken commented on src/api/app/models/project/update_from_xml_command.rb:
```diff
@@ -1,6 +1,6 @@
 class Project
   class UpdateFromXmlCommand
-    include Project::Exceptions
```
Same here, please squash it to the commit that introduced this. This also makes it easier to review the PR

vpereira commented:
@adrianschroeter looks like https://github.com/openSUSE/open-build-service/commit/e7cc3731ff02d71a966596c9c440a4d830aa729c broke minitest.. can we rever this commit or there is a follow-up commit coming?

vpereira commented:
![geeko-catwalk](https://i.gifer.com/3FAy.gif)

dmarcoux commented on src/api/app/controllers/application_controller.rb:
```diff
@@ -213,7 +213,7 @@ def pass_to_backend(path = nil)
   end
 
-  rescue_from APIException do |exception|
+  rescue_from APIError do |exception|
```
Now it's weird with `APIError`

coolo replied:
Classic case of 'Kann man machen, muss man aber nicht'

coolo commented on src/api/app/models/service.rb:
```diff
@@ -1,7 +1,6 @@
-class Service
-  include ActiveModel::Model
-  class InvalidParameter < APIException; end
+class Service < ActiveXML::Node
```
why revert that?

vpereira commented:
@Ana06 please stop to link others contributions guide. We have our own and we follow just our own. The PR was already approved by @bgeuken. Please anyone approve it @openSUSE/open-build-service ?

hennevogel commented:
well @vpereira our own [contributing guide](https://github.com/openSUSE/open-build-service/blob/master/CONTRIBUTING.md#proper-commit-messages) requests similar things. 

In f60d62cd you just repeat what you did instead of telling us **why**. 

And reviews are not a 0 sum game. One negative review is enough 😇

vpereira commented:
@hennevogel what was exactly the "negative review"? I couldn't find any.
